cli/cliui/select.go

```diff
@@ -114,35 +253,191 @@ func MultiSelect(inv *serpent.Invocation, opts MultiSelectOptions) ([]string, er
 		return opts.Defaults, nil
 	}
 
-	prompt := &survey.MultiSelect{
-		Options: opts.Options,
-		Default: opts.Defaults,
-		Message: opts.Message,
+	options := make([]*multiSelectOption, len(opts.Options))
+	for i, option := range opts.Options {
+		chosen := false
+		for _, d := range opts.Defaults {
+			if option == d {
+				chosen = true
+			}
+		}
+
+		options[i] = &multiSelectOption{
+			option: option,
+			chosen: chosen,
+		}
 	}
 
-	var values []string
-	err := survey.AskOne(prompt, &values, survey.WithStdio(fileReadWriter{
-		Reader: inv.Stdin,
-	}, fileReadWriter{
-		Writer: inv.Stdout,
-	}, inv.Stdout))
-	if errors.Is(err, terminal.InterruptErr) {
-		return nil, Canceled
+	initialModel := multiSelectModel{
+		search:  textinput.New(),
+		options: options,
+		message: opts.Message,
+	}
+
+	initialModel.search.Prompt = ""
+	initialModel.search.Focus()
+
+	m, err := tea.NewProgram(
+		initialModel,
+		tea.WithInput(inv.Stdin),
+		tea.WithOutput(inv.Stdout),
+	).Run()
+
+	values := []string{}
+	if m, ok := m.(multiSelectModel); ok {
+		if m.canceled {
+			return values, Canceled
+		}
+
+		values = m.selectedOptions()
 	}
 	return values, err
 }
 
-type fileReadWriter struct {
-	io.Reader
-	io.Writer
+type multiSelectOption struct {
+	option string
+	chosen bool
+}
+
+type multiSelectModel struct {
+	search   textinput.Model
+	options  []*multiSelectOption
+	cursor   int
+	message  string
+	canceled bool
+	selected bool
+}
+
+func (multiSelectModel) Init() tea.Cmd {
+	return nil
+}
+
+//nolint:revive // For same reason as previous Update definition
+func (m multiSelectModel) Update(msg tea.Msg) (tea.Model, tea.Cmd) {
+	var cmd tea.Cmd
+
+	if msg, ok := msg.(tea.KeyMsg); ok {
```

Comment:
nit: invert and early return

Reply:
I've slightly changed the logic here. I just noticed it is slightly different to how the `func (m selectModel) Update` function works. `m.search.Update` can handle more than just a `tea.KeyMsg`. With the updated logic does this still make sense?